_source/_posts/2023-07-27-spring-graphql-react.md

```diff
@@ -0,0 +1,1794 @@
+```
+
+Add the client secret to the `.env` file:
+
+__.env__
+```shell
+export OKTA_OAUTH2_CLIENT_SECRET={clientSecret}
+```
+
+Add the following factory method to the class `SpringBootGraphQlApiConfig`, for requiring a bearer token for all requests:
+
+__SpringBootGraphQlApiConfig.java__
+```java
+    ...
+    @Bean
+    public SecurityFilterChain configure(HttpSecurity http) throws Exception {
+        http.oauth2ResourceServer(oauth2ResourceServer -> oauth2ResourceServer.jwt(withDefaults()));
+        return http.build();
+    }
+    ...
+```
+
+> __NOTE:__
+> The Okta Spring Boot starter provides the security auto-configuration out of the box, and the resource server configuration would not be necessary. For some reason, the Spring for GraphQL cors allowed origins configuration does not take effect without the customization above.
+
+Again, in the root folder, run the API server with:
+
+```shell
+source .env && ./gradlew bootRun
+```
+
+With HTTPie, send a request to the API server using a bearer access token:
+
+```shell
+ACCESS_TOKEN={auth0AccessToken}
+```
+
+```shell
+echo -E '{"query":"{\n    companyList(page: 20) {\n        id\n        SIC\n        name\n        status\n        category\n        companyNumber\n        countryOfOrigin\n    }\n}"}' | \
+  http -A bearer -a $ACCESS_TOKEN POST http://localhost:8080/graphql
+```
+
+> __NOTE:__
+> Follow [these instructions](https://auth0.com/docs/secure/tokens/access-tokens/get-management-api-access-tokens-for-testing) to get an access token for the Auth0 Management API, which can be used for testing the server API.
+
+### Add Auth0 Login to the React client
+
+When using Auth0 as the identity provider, you can configure the Universal Login Page for a quick integration without having to build the login forms. First, create a SPA application using the Auth0 CLI:
+
+```shell
+auth0 apps create \
+  --name "React client for GraphQL" \
+  --description "SPA React client for a Spring GraphQL API" \
+  --type spa \
+  --callbacks http://localhost:3000/callback \
+  --logout-urls http://localhost:3000 \
+  --reveal-secrets
+```
+
+Copy the Auth0 domain and the client ID, and update the `src/.env.local` adding the following properties (add the new variables to the example file, too):
+
+__.env.local__
+```shell
+NEXT_PUBLIC_AUTH0_DOMAIN={yourAuth0Domain}
+NEXT_PUBLIC_AUTH0_CLIENT_ID={clientId}
+NEXT_PUBLIC_AUTH0_CALLBACK_URL=http://localhost:3000/callback
+NEXT_PUBLIC_AUTH0_AUDIENCE=https://{yourAuth0Domain}/api/v2/
+```
+
+For handling the Auth0 post-login behavior, you need to add the page `src/app/callback/page.tsx` with the following content.
+
+__callback/page.tsx__
+```tsx
+import Loader from "@/components/loader/Loader";
+
+const Page = () => {
+  return <Loader/>
+};
+
+export default Page;
+```
+
+For this example, the callback page will render empty.
+
+Add the `@auth0/auth0-react` dependency to the project:
+
+```shell
+npm install @auth0/auth0-react
```

Comment:
Why aren't you using [nextjs-auth0](https://github.com/auth0/nextjs-auth0) here?

Reply:
Added the note.